### ISSUE TITLE
Soft-Q-Learning RL Support

### ISSUE THREAD

**Issue by amarazad** [huggingface/trl#443]
Hi @lvwerra, 
Do you have any plans to implement the Soft-Q-Learning algorithm  [[paper](https://aclanthology.org/2022.findings-emnlp.518.pdf)] as well. The algorithm has shown better results then PPO for various text generation tasks especially for low good data scenarios and also used in recent prompt learning task such as [rl prompt ](https://arxiv.org/pdf/2205.12548.pdf). It would be really nice to have this algorithm implemented under the same framework.

**Comment by lvwerra** [comment]
Hi @amarazad currently nobody is working on this as far as I know but we are open to contributions if you want to give it a shot!

**Comment by amarazad** [comment]
Hi @metric-space, Awesome, it will be great. Please go ahead. 
I might not have time immediately but would love to help latter if needed. I will be the first user for sure. :)

**Comment by amarazad** [comment]
@metric-space For quick reference, one version is implemented [here](https://github.com/HanGuo97/soft-Q-learning-for-text-generation). It is also implemented for RL prompt [here](https://github.com/mingkaid/rl-prompt).

**Comment by metric-space** [comment]
@amarazad re:first-user that will be excellent. Thank you for the references. Will be starting work on this promptly :+1:

**Comment by metric-space** [comment]
Update on this: still reading through the referenced papers and code and figuring out how everything lines up. Hopefully will have a more solid update either codewise or a question or both

**Comment by metric-space** [comment]
Update on this: Apologies everyone especially @amarazad  I'm not able to deliver a first draft as I am still working through this.
 
That said, I can talk about what exactly I am trying to achieve here:
1. I am trying my best to make sure not to introduce any new base classes like the one here https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/main/sql/modules_base.py
 I'm trying to merge the logic here (https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/main/sql/modules.py#L19) and here (https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/main/experiments/translation.py)  into a trainer class akin to the `ppo_trainer.py` file
 Took me some time to figure out if I can just work with the existing transformer classes rather than reintroduce this as the authors of the paper (soft-q-learning) did here: https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/main/modules/models.py#L32

2. Something I've been exploring, it seems like the model being used for RLPrompt  i.e the lm-adaptor (https://github.com/mingkaid/rl-prompt/blob/main/rlprompt/models/lm_adaptor_model.py) is something that should exist in the `PEFT` library. One major reason for doing so because it would be a natural addition to the ecosystem and calling it within the `trl` library as a peft model with the correct type seems much more nicer  than trying to re-introduce something  adaptor'ish here(`trl`) 

I'm not committing to any future date but am actively working on this. Figuring out what i aligns ; the finer details of the existing implementations and maintaining the same style is taking time. I recognize I have overwhelmed myself a bit but nothing I can't handle given more time. Any suggestions and thoughts are welcome especially if it sems like I may complicating things 

P.S still figuring the finer details of naturally incorporating RLPrompt using the future `SoftQLearningTrainer` (+ `SoftQLearningConfig`) and (if considered sensible) the adapter LM model as `PEFT` class

**Comment by amarazad** [comment]
@metric-space I guess its going well. I like your approach; i.e.  not to introduce any new base classes and taking the lm-adapter to PEFT lib. BTW, nothing to apologies, its perfectly fine. Looking forward to it.

**Comment by metric-space** [comment]
Hey folks,

Still working through the first draft. Before you read the following, this isn't exactly a show stopper but something I thought might be worth asking 

One of the things that is adding to the complexities in implementation is the transfer from using a transformer model implemented using this library [texar](https://github.com/asyml/texar-pytorch) to using transformer models defined in the [transformers]() library. And of course, chasing small details here and there

A key aspect of the implementation of the the [paper behind all of this](https://aclanthology.org/2022.findings-emnlp.518.pd) is the connection between `logits` and Q-values. 

A question arose trying to get the transformer library's equivalent of this particular line in the reference implementation: 
https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/93e8913fa141930c632ccc53eacaed98a46977cc/sql/modules.py#L213 [a]
This happens to be a generation call (that involves sampling) that returns output logits and the sampled ids (using the `texar` library)
There is a straightforward translation/mapping of this call using huggingface's transformer library when it comes to the sampled ids but as for the logits part, this isn't as straightforward. There exists something called `scores` that will be returned as part of the library call

```python
self.model.generate(
                batch[0].input_ids,
                do_sample=True,
                top_k=self._top_k,
                top_p=self._top_p,
                return_dict_in_generate=True,
                output_scores=True
            )

```
however there is enough evidence that the `scores`  are not the same as the `logits`
https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2652-L2661 [b]
As per the code in the above link, the `logits` become tampered with (which is expected, as scores have to with the generation aspect)

seems like the `texar` equivalent that is being used gives back untampered logits as evidenced here: https://github.com/asyml/texar-pytorch/blob/5b56ad39ff980c7a2129d59158d43c809c4b43db/texar/torch/modules/decoders/transformer_decoders.py#L763-L764 [c]

**Questions**

1. My question for @amarazad, this I take is a difficult question but I have to ask: would the difference between logits and scores matter when it comes to getting results? planning to try this out but was curious about a  take on this theoretically 
2. My question for @lvwerra and @younesbelkada , would you folks be open to accepting a PR (for the transformers library) to get the logits defined here https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2652 into the output alongside scores ? It seems beyond my use case there exists some utility as evidenced by https://github.com/asyml/texar-pytorch/blob/5b56ad39ff980c7a2129d59158d43c809c4b43db/texar/torch/modules/decoders/transformer_decoders.py#L763-L764 

**Traces**
[a]  
 - https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/93e8913fa141930c632ccc53eacaed98a46977cc/sql/modules.py#L213
 - https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/93e8913fa141930c632ccc53eacaed98a46977cc/modules/models.py#L278
 - https://github.com/HanGuo97/soft-Q-learning-for-text-generation/blob/93e8913fa141930c632ccc53eacaed98a46977cc/modules/models.py#L186
 
[b]
- https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L1024
- https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L471
- https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L1161
- https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L2652-L2661

**Comment by metric-space** [comment]
Just an update: my progress off late on this has been pretty slow, will be picking up steam on this within the next two to three days. I have not forgotten about this and will be seeing this to the end

**Comment by amarazad** [comment]
> My question for @amarazad, this I take is a difficult question but I have to ask: would the difference between logits and scores matter when it comes to getting results? planning to try this out but was curious about a take on this theoretically

@metric-space somehow I slipped over the question tagged to me; my apologies. I think it should not matter from result perspective whether we take the logits or scores. However, handle on logits will allow to control /modify the action distribution if needed latter on.  So, it depends, if its much of a hassle we should keep logits as well.

**Comment by metric-space** [comment]
@amarazad Thank you for the reply and no worries at all. So I do have a bit of an update . So the current draft is in a working state. Code is still very shabby and needs more work, I haven't put in a proper tracker like tensorboard/wandb for metric tracking. But the code works. The losses associated are extremely bad so I have to look through where exactly I am going wrong, at the moment this is what I am working on

If you're interested in seeing for yourself:
1.  pull down the work https://github.com/lvwerra/trl/pull/510
2.  go to the `examples/soft-q-learning` folder
3.  `bash download_demo_data.sh`
4.  `python3 soft_q_learning.py`

The metrics as of now are logged to standard output
Expect more updates within the next few days

**Comment by metric-space** [comment]
Update: seems like the incredulously high loss numbers were caused by 
1. unnecessary target model params addition (my fault entirely)
2. weird character presence in training data and a seemingly unnecessary tokenization step in the reward function (still looking through but removed them for now)

the loss numbers seem on par with the original implementation but not quite there, still looking through

**Comment by metric-space** [comment]
Don't plan on abandoning this given the work put in. Just busy at the moment

**Comment by metric-space** [comment]
Re-commencing work on this starting tomorrow

**Comment by amarazad** [comment]
@metric-space My colleague, @Brij-Chavda, has tried using the softQLearning and has some question. @Brij-Chavda Can you please post questions.

**Comment by Brij-Chavda** [comment]
Hi @metric-space, thanks for integrating soft-q-learning. I have a couple of questions, 
1. Can we put the GPT2 model as policy LM with only an adaptor trainable layer?
2. Can we use the current reward function for classification objectives as well? (or use other reward functions)

**Comment by metric-space** [comment]
@Brij-Chavda Hey there, and you're welcome and thank you for your patience

Before I answer your questions, take note the code is still pretty raw. I'm still getting back into the swing of things when it comes to it's development. I think there is a lot more cleaning up to do from my end

1.  The original code and it's current integrated form (as far as I remember) only dealt with encode-decoder models. I haven't run the code with a decoder only model like gpt-2. I'll be running it with a `gpt-2` model as the policy model wither today or tomorrow.  I think it shouldn't be a problem and if it is, it will be a difference in the api and I should be able to abstract it out as a separate standard class placing the burden on the user to map the model. 
As for the adaptor trainable layer, I reason because of this line here https://github.com/huggingface/trl/blob/6695341fdb32092693b5e12936a90e636dae4c8f/trl/trainer/soft_q_learning_trainer.py#L68 it should work. 

2. > Can we use the current reward function for classification objectives as well? (or use other reward functions)

I think yes, I think the current state of the code makes this less obvious. So while I believe the answer is yes, I would wait for scheduled cleanup

**Comment by metric-space** [comment]
Update: gpt2 doesn't work right off the bat

**Comment by amarazad** [comment]
@metric-space First of all thanks a lot for keep this going. 
Re: gpt2.  For encoder-decoder (T5) and decoder (GPT2) model the labels and attention mask has to be different. 
 I think it should be good if you can abstract it out and let the user map it. Latter it can be extended to support each one case.

**Comment by metric-space** [comment]
There has been activity on the associated PR

**Comment by metric-space** [comment]
Still planning to do this

**Comment by lvwerra** [comment]
stalebot vs metric-space, the epic battle 😄

**Comment by metric-space** [comment]
please keep this open
